### ISSUE TITLE
Cache Bitmap OCR results

### ISSUE THREAD

**Issue by sarav** [farkam135/GoIV#141]
I gave it a shot in #140 . The logic works.

If the Bitmap is fixed, I don't see any reason the OCR results should change each time. I'd also think that the OCR code is a lot more complicated and floating point intensive than hashcode generation. Unfortunately, Bitmap.hashcode() doesn't generate a hashcode based on the bitmap picture. So, I had to extract the pixels to do it. I'm unable to officially verify it though. I guess I could do a timestamp check. Didn't think of that -- but I got to log off for today.

Can others (@Blaisorblade @Eccenux @mattaereal @nahojjjen @farkam135) give it a shot and let us know what you think? Does it feel a bit faster when you check the IV for the same type of Pokemon (assuming you didn't rename it) or when you check the same Pokemon again? Is the LRU size of 200 too small? I'd like to hear some thoughts.

**Comment by farkam135** [comment]
It might just be me but I usually don't rescan unless there was an obstruction. It sort of ruins the opportunity for me to rescan it at a better time if the results are cached.

**Comment by sarav** [comment]
I think you misunderstood the issue/code (not sure you looked at it). If the issue is an obstruction, the hashes won't match. I'm using the cached value only if the "pictures" are identical. If the pictures are identical, I'd expect tesseract to give the same result.

I rescan because I go back and forth between a few pokemon that I'm trying to decide on transferring and I keep forgetting their IV range.

**Comment by farkam135** [comment]
Ah yea, just looked at the code my bad, how much size does each cached scan take up?

**Comment by agrajaghh** [comment]
Do you think you will ever get two identical bitmaps? Since there is the background animation with the bubbles and the pokemon is moving..

**Comment by Blaisorblade** [comment]
> I'm using the cached value only if the "pictures" are identical.

Well... if there are no collisions. Avoiding this problem (with a `HashMap<HashableImage, String>` from a new `HashableImage` class with a suitable `hashCode` method), OTOH, would mean storing the image.

**Comment by Blaisorblade** [comment]
> Do you think you will ever get two identical bitmaps?

The patch hashes the areas containing the pokemon and candy names, so that shouldn't matter.

**Comment by agrajaghh** [comment]
> The patch hashes the areas containing the pokemon and candy names, so that shouldn't matter.

ah okay, so hashing would just be useless for CP...

**Comment by Blaisorblade** [comment]
Re size: right now we use little because we don't store the images.

Each LRU cache entry stores the result string (say 8 char) and a "hash" made of <=8 characters (16 bytes), which can go down to 4, plus overhead, and this gives a size bound of <= 16 chars * 2 bytes/char * 200 entries = 6.4KB overall.

On a 1080p phone, instead, the name area uses ~1920/18 * 1080/2 = 107 * 960 pixels, and at 32 bits per pixel that would give ~410K per image, or 82M for 200 entries. IOW, this is not feasible without compression.

**Comment by sarav** [comment]
@Blaisorblade 
> Well... if there are no collisions. Avoiding this problem (with a HashMap<HashableImage, String> from a new HashableImage class with a suitable hashCode method), OTOH, would mean storing the image.

Which is why I didn't do it. It'll bloat up really quick. Looks like you got to the same conclusion. Do you think the hashing is faster than OCR? I think it should be, but I didn't verify.

@agrajaghh If you see the code, I don't cache the CP stuff as it changes too often to be likely to get a hit and the hash calculation might do more work than it saves for CP.

**Comment by Blaisorblade** [comment]
> Do you think the hashing is faster than OCR? I think it should be, but I didn't verify.

I'm assuming so. My main concern is that because of hash collisions we might get false OCRs results.
I feared that was a showstopper, but since this is used as a heuristic detection, if the collisions are seldom enough this might be just fine. I'll try to test your patch.

**Comment by sarav** [comment]
Yeah, hash collisions are my concern too and same reasoning I had on why it might be fine. Thanks for testing it out.

**Comment by agrajaghh** [comment]
regarding the collision, I think thats safe:

> Probability of just two hashes accidentally colliding is 1/2^128 which is 1 in 340 undecillion 282 decillion 366 nonillion 920 octillion 938 septillion 463 sextillion 463 quintillion 374 quadrillion 607 trillion 431 billion 768 million 211 thousand 456.

> However if you keep all the hashes then the probability is a bit higher thanks to birthday paradox. To have a 50% chance of any hash colliding with any other hash you need 2^64 hashes. This means that to get a collision, on average, you'll need to hash 6 billion files per second for 100 years.

http://stackoverflow.com/questions/201705/how-many-random-elements-before-md5-produces-collisions

**Comment by sarav** [comment]
But do you know md5 is the algorithm that Arrays.hash is using? There was this other thread I saw where storing the English dictionary caused about a 100 collision.

But I agree, the likelihood is low.

**Comment by agrajaghh** [comment]
oh you're right its just a int (32bit), I don't know why I had md5 in my mind...

If you scroll down a bit there is a good graph about the probability of collisions:
http://eclipsesource.com/blogs/2012/09/04/the-3-things-you-should-know-about-hashcode/ 

also:
> Whenever it is invoked on the same object more than once during an execution of a Java application, the hashCode method must consistently return the same integer, provided no information used in equals comparisons on the object is modified. This integer **need not remain consistent from one execution of an application to another** execution of the same application.

So it would be a lot safer with sha1 or even md5, but for that you probably have to convert the bitmap to a byte array or something like that...

**Comment by Blaisorblade** [comment]
In general: most hashes are not cryptographically robust to be (much) faster, and also because collisions usually just cause a few extra steps in hashtable lookup (the hash has to be decent, but not cryptographically strong).

Here the goal is saving CPU even at the cost of slightly less accuracy.
Maybe MD5 is still slower than OCR, but not sure.

**Comment by sarav** [comment]
So, are we in agreement that I should try to merge this now? Did anyone else test it or profile it?

**Comment by Blaisorblade** [comment]
Haven't been able to test the PR, but the idea is fine and OCR does take time.

**Comment by sarav** [comment]
Oh god! I ran method tracing to see how well OCR caching worked. I was totally right about not doing the hashing after replaceColors. I did one pokemon scan and saw how long it took.


```
Function name     Calls    Total Time (ms)	Time (ms) just in fn (excluding child calls)
hashBitmap          3          0009.562                 0000.3 ms
replaceColors       4          2676.000                 1566.0 ms
```

Yes, that 2.6 SECONDS to run replaceColors vs 9 milliseconds to run the hashBitmap. This was done before the patch to remove floating point operations you created due to this discussion. But even if you assume all the code that replaced the floating point code is zero, it still looks like about 1 second of run time vs 9ms. About a 100x improvement if we get an OCR cache hit!

I'm convinced on the OCR cache at this point.

**Comment by sarav** [comment]
Btw, the Tess APIs don't seem to take a lot of time in the method trace, but that could be due to the native code they are using. Even in that case, I'd expect it to be accounted properly under total time, but maybe not.

So, replaceColors is the major culprit! I'll do one or two more runs to check this and then head to sleep. I'm already up too late for tomorrow.

**Comment by Blaisorblade** [comment]
That's why they say performance is hard to predict and profiling is always good. 😄 

Still, it'd be good to see how much is left after the patch. And if that's still so expensive we *might* want to move `replaceColors` to the GPU (I can't do that, but this task should be much faster and more power-efficient on a GPU; not sure on the setup overhead though).

> Btw, the Tess APIs don't seem to take a lot of time in the method trace, but that could be due to the native code they are using. Even in that case, I'd expect it to be accounted properly under total time, but maybe not.

As in, maybe OCR is negligible compared to `replaceColors`? Interesting.

> method tracing

That's a tracing profiler instead of a sampling one? Those are also useful but they add much more overhead... [Docs say don't trust this](https://developer.android.com/studio/profile/am-methodtrace.html): 

> Running the method trace significantly affects CPU timings. Use the method trace to understand the flow of the program, but not for performance timings.

They don't give alternatives—maybe one can trace by hand those methods (by logging timestamp differences from begin and end).

**Comment by sarav** [comment]
@Blaisorblade I think you missed my edit. Even after removing the FP code it takes 1.5 seconds or so. I checked a few times.

I was thinking about moving this to the GPU too and looking stuff up. But the code doesn't seem obvious. Also, this might not just about the GPU vs CPU (I'm sure GPU still beats CPU), but also how we access the data and the cache misses (the actual HW cache), etc.

As for tracing, yeah, I know it's not accurate, but it should still be proportional I think. And we know that it takes a while. So, it's not like tracing it making it that much slower.

**Comment by Blaisorblade** [comment]
I had missed the edit, thanks. I just found sampling profilers here: https://developer.android.com/studio/profile/ddms.html#profiling
Feel free to leave it alone, but I might just try it 😄 

And in any case, thanks for the find and 👍  for the caching.

**Comment by sarav** [comment]
Method profiling is exactly what I used. Anyway, I'm done for the day. I have more ideas on this.

Also, does someone have the history on why replaceColors was added? How much did it help? Why do we need the distance/similarity check instead of just doing a color comparison. I'd this that's not much color variance in within the text. So, can't we just kill rest of the colors?

@farkam135 @nahojjjen

**Comment by nahojjjen** [comment]
The CP scan needs color replacing, because the background changes and sometimes is colorful enough to disrupts the tesseract. It can however maybe be changed to just keeping only the cp text color - Ill have to experiment and see how much it changes the reading.

**Comment by Blaisorblade** [comment]
@nahojjjen But why do the non-CP readings all use replaceColor? That's all black/grey on white no?

@sarav 
> Method profiling is exactly what I used.

Trace-based profiling or sample-based profiling? They're very different.

**Comment by sarav** [comment]
Also, CP, I totally expect and understand the need. But what about name, candy name and HP. They seem clean enough already.

**Comment by sarav** [comment]
@Blaisorblade how do I check which one I used? The one I used generated .trace files, but it could still be sample based. Don't know yet.

**Comment by sarav** [comment]
Btw, I've been running the OfflineDebug version of the apk for all my testing. I tried the Offline Release one and it feels much snappier. Not sure if it's the placebo effect or if there's significant difference. But I'm running the one without FP code. So, there's that.

**Comment by sarav** [comment]
Definitely placebo. I didn't even install it *facepalm* I need sleep.

**Comment by agrajaghh** [comment]
I just tried meassuring it with [TimingLogger](https://developer.android.com/reference/android/util/TimingLogger.html) and it looks much faster: (Sony Xperia Z3 Compact)
```
replaceColors start: begin
replaceColors start:      21 ms,
replaceColors start: end, 21 ms
replaceColors start: begin
replaceColors start:      10 ms,
replaceColors start: end, 10 ms
replaceColors start: begin
replaceColors start:      6 ms, 
replaceColors start: end, 6 ms
replaceColors start: begin
replaceColors start:      9 ms, 
replaceColors start: end, 9 ms
```

I added this at the beginning of replaceColors:
```
TimingLogger timings = new TimingLogger("MY_TAG", "replaceColors start");
```
and this at the end:
```
timings.addSplit("");
timings.dumpToLog();
```
you also have to run `setprop log.tag.MY_TAG VERBOSE` in adb shell

**Comment by agrajaghh** [comment]
just for completeness the timings for the OCR:
```
`name OCR start: begin
name OCR start:      261 ms, 
name OCR start: end, 261 ms
Candy OCR start: begin
Candy OCR start:      248 ms,
Candy OCR start: end, 248 ms
HP OCR start: begin
HP OCR start:      58 ms, 
HP OCR start: end, 58 ms
CP  OCR start: begin
CP  OCR start:      111 ms, 
CP  OCR start: end, 111 ms
```

**Comment by Blaisorblade** [comment]
@agrajaghh Thanks! Do you have a branch with your changes? I didn't find your fork :-)
And I guess you were not using my changes in https://github.com/farkam135/GoIV/pull/158, right?

**Comment by sarav** [comment]
Can you please time the hashBitmap function too?

**Comment by agrajaghh** [comment]
@sarav I just timed `myBitmap.hashCode();` at the beginning of `replaceColors()` and I'm always getting 0ms. here is the branch: https://github.com/agrajaghh/GoIV/tree/OCR_timing

@Blaisorblade no, just plain master, but with your PR it looks like that:
```
replaceColors: begin
replaceColors:      0 ms, myBitmap.hashCode()
replaceColors:      11 ms, 
replaceColors: end, 11 ms
replaceColors: begin
replaceColors:      0 ms, myBitmap.hashCode()
replaceColors:      3 ms, 
replaceColors: end, 3 ms
replaceColors: begin
replaceColors:      0 ms, myBitmap.hashCode()
replaceColors:      2 ms, 
replaceColors: end, 2 ms
replaceColors: begin
replaceColors:      0 ms, myBitmap.hashCode()
replaceColors:      3 ms, 
replaceColors: end, 3 ms
```
so over 50% savings :)
https://github.com/agrajaghh/GoIV/tree/OCR_timing_PR158

**Comment by sarav** [comment]
hashBitmap() is different from bitmap.hashcode()

**Comment by agrajaghh** [comment]
Damn, you're right. I was on the run and did it in a hurry :S Damn auto complete... I will use hashBitmap when I'm back home...

**Comment by Blaisorblade** [comment]
@agrajaghh your branch IIUC also doesn't have `hashBitmap`, you need @sarav's PR. `bitmap.hashCode()` IIUC is fast since it ignores the pixels 😉 .

Meanwhile: with the current evidence I'm in favor of @sarav's caching — `replaceColors` and `hashBitmap` are basically a loop over all pixels and should have comparable cost, while OCR takes much longer. (No strong opinion on the order of `replaceColors` vs `hashBitmap`, you needn't convince me).

I currently expect the original timing for `replaceColors` were bigger because of trace-based profiling (which adds instrumentation to each method call), though the difference is bigger than I feared. Sampling-based profiling has much lower overhead when it works.

**Comment by agrajaghh** [comment]
okay, I hope I got it right now:
https://github.com/agrajaghh/GoIV/tree/OCR_timing_hash

I was running `hashBitmap()` from #140 (at least I think so, because @sarav deleted that branch, so I had to take it from the outdated comments) and MD5 hash on the bitmap each 100 times. hashBitmap() is 10% to 50% faster, but both are very fast (1-4 ms) in comparison to the OCR which is taking > 100 ms

```
310 ms, 100x hashBitmap()
392 ms, 100x MD5

184 ms, 100x hashBitmap()
347 ms, 100x MD5

112 ms, 100x hashBitmap()
165 ms, 100x MD5

220 ms, 100x hashBitmap()
237 ms, 100x MD5
```

**Comment by sarav** [comment]
Thanks. Looked at your commit. The hashBitmap looks fine. This is some good data assuming the timing code APIs are used correctly. 

I think it justifies both the floating point removal patch and the OCR caching patch.

**Comment by Blaisorblade** [comment]
Indeed, I'm happy with both ones. Can you prepare an updated PR? I think this is not controversial 👍

**Comment by Blaisorblade** [comment]
I've renamed the issue, since IMHO we actually decided to do this now.

**Comment by sarav** [comment]
Cool. I'll push it soon.

**Comment by sarav** [comment]
@Blaisorblade I've joined the #dev channel but I don't see any chats and my messages are not showing up when I send.

**Comment by Blaisorblade** [comment]
And BTW: thanks a lot @agrajaghh for the benchmarking work.

**Comment by sarav** [comment]
Created PR #178

**Comment by Blaisorblade** [comment]
PR merged.

BTW next time, you @sarav can write "Fixes #141" in the PR or commit message body, so merging the PR doesn't forget the issue open. I could have noticed.
